core/src/main/java/org/apache/iceberg/StaticTableScan.java

```diff
@@ -25,16 +25,33 @@
 
 class StaticTableScan extends BaseTableScan {
   private final Function<StaticTableScan, DataTask> buildTask;
+  // Metadata table name that the buildTask that this StaticTableScan will return data for.
+  private final String scannedTableName;
 
-  StaticTableScan(TableOperations ops, Table table, Schema schema, Function<StaticTableScan, DataTask> buildTask) {
+  StaticTableScan(TableOperations ops, Table table, Schema schema,  String scannedTableName,
+                  Function<StaticTableScan, DataTask> buildTask) {
     super(ops, table, schema);
     this.buildTask = buildTask;
+    this.scannedTableName = scannedTableName;
   }
 
-  private StaticTableScan(TableOperations ops, Table table, Schema schema,
+  private StaticTableScan(TableOperations ops, Table table, Schema schema,  String scannedTableName,
                           Function<StaticTableScan, DataTask> buildTask, TableScanContext context) {
     super(ops, table, schema, context);
     this.buildTask = buildTask;
+    this.scannedTableName = scannedTableName;
+  }
+
+  @Override
+  public TableScan appendsBetween(long fromSnapshotId, long toSnapshotId) {
+    throw new UnsupportedOperationException(
+        String.format("Incremental scan is not supported for metadata table %s", scannedTableName));
+  }
+
+  @Override
+  public TableScan appendsAfter(long fromSnapshotId) {
+    throw new UnsupportedOperationException(
+        String.format("Incremental scan is not supported for metadata table %s", scannedTableName));
```

Comment:
Should this PR also add a similar update to `FilesTableScan`? It also looks like we may need to update `BaseAllMetadataTableScan`. I see that `AllManifestsTableScan` overrides `useSnapshot` and `asOfTime` but not the incremental methods that also don't work.

Reply:
I think so. Same thing for `EntriesTableScan`.

Currently, `FilesTableScan` and `EntriesTableScan` inherit directly from `BaseTableScan`. However, they could easily be made to inherit from `StaticTableScan` and then just override the methods that they need.

I wanted to keep this PR small so that it would be easy to digest, but I think that if we refactor, we could easily make `StaticTableScan` into the base class for all metadata table scans other than the aggregate metadata table scans (the ones that derive from `BaseAllMetadataTableScan`) - essentially making it into `BaseMetadataTableScan` to go along with `BaseAllMetadataTableScan`.

Currently, `FilesTableScan` and `EntriesTableScan` inherit directly from `BaseTableScan`. However, they could easily be made to inherit from `StaticTableScan` and then just override the methods that they need. This way, the overrides like `useSnapshot` and `asOfTime` which are currently handled in a few places could all be handled in one place.

Then, if we chose to implement that functionality for a particular metadata table type, we could just override in the actual implementing scan class.

I can update this PR so that `FilesTableScan` and `EntiresTableScan` inherit from `StaticTableScan`, or I can add a similar update to `FilesTableScan` (and possibly `EntriesTableScan`) or handle them in a follow up PR.

Reply:
**TLDR**

It seems to me that the inheritance is mostly because of the time at which some of this code was written, but I believe it could (and should) be unified such that for each `MetadataTableType`:

The current layout of refined error messages for time based / snapshot specific / incremental metadata table scans:
```
MetadataTableType  Current Scan Class Super Class         Refined Time Based Scan Specific Overrides
  ENTRIES                       BaseTableScan                                       None
  FILES                             BaseTableScan                                       None
  HISTORY                       StaticTableScan                                     None
  SNAPSHOTS                StaticTableScan                                      None
  MANIFESTS                  StaticTableScan                                     None
  PARTITIONS                 StaticTableScan                                     None
  ALL_DATA_FILES        BaseAllMetadataTableScan                   [asOfTime, useSnapshot](https://github.com/apache/iceberg/blob/32ce969218b75915cb7396c541bdc19f4ac8bd9f/core/src/main/java/org/apache/iceberg/AllDataFilesTable.java#L92-L100)
  ALL_MANIFESTS         BaseAllMetadataTableScan                  [asOfTime, useSnapshot](https://github.com/apache/iceberg/blob/32ce969218b75915cb7396c541bdc19f4ac8bd9f/core/src/main/java/org/apache/iceberg/AllManifestsTable.java#L99-L107)
  ALL_ENTRIES               BaseAllMetadataTableScan                  None
  ```

All of the above that have overrides are implemented in the final implementing class, as shown in the links.

I think it would be much better to pass the actually scanned table name (and not just the table object, which references the Iceberg table and not its metadata table) to the two base scan classes, and then have consistent error messages for all which state the table name attempting to be scanned for users (as users encounter this problem often enough when switching back and forth between editing a snapshot specific query and inspecting a metadata table).

So I propose:
```
MetadataTableType            New Scan Class
  ENTRIES               **EntriesTableScan extends `StaticTableScan**`                                  None
  FILES                 **FilesTableScan extends StaticTableScan**                                 None
  HISTORY                 StaticTableScan                                     None
  SNAPSHOTS               StaticTableScan                                      None
  MANIFESTS               StaticTableScan                                     None
  PARTITIONS              StaticTableScan                                     None
  ALL_DATA_FILES        BaseAllMetadataTableScan                   _None -> Moved to BaseAllMetadataTableScan_
  ALL_MANIFESTS         BaseAllMetadataTableScan                  _None -> Moved to BaseAllMetadataTableScan_
  ALL_ENTRIES               BaseAllMetadataTableScan                  None
  ```



If we decide to implement the functionality, then the final class can override it. This would cut down on code duplication and, imho, make the code much more readable.

Passing the additional metadata (actual table name, as well as the metadata table type ideally) would not affect the serialization done on static table scans, as the information is only used when building the query (e.g. prior to distributing the information across the cluster for spark and likey all engines).

Reply:
I can update this PR so that `FilesTableScan` and `EntiresTableScan` inherit from `StaticTableScan`, or I can add a specific override without touching the inheritance further (or at all in this PR) for the ones you mentioned, or handle some of it in a follow up PR.

I know that's a lot, but I've looked over it quite a bit and I think this would make things a lot cleaner. There's a decent amount of code that could be refactored and slimmed down to be more DRY.

Reply:
Or if we prefer, I can just keep the structure as is for the other ones and override methods there.

Reply:
I implemented the refined error message in all metadata table classes for `appendsBetween` and `appendsAfter`, and updated the test to test every single metadata table type.

I did not know if I should go for `asOfTime` and `useSnapshot`, though I could pick those up too if you think that's a good idea @rdblue.

I've also uncovered a bug in the emitted `ScanEvent` that will be very easy to fix once this is merged: https://github.com/apache/iceberg/issues/2635

Reply:
I hit all of the existing places you mentioned @rdblue. I'm going to resolve this conversation.

I chose not to refactor the current scan classes that extend `BaseScanClass` to use `StaticScanClass` in hopes of making this PR more digestible.

I can do that in a follow up if needed be.

But this now passes the unit test for the refined error message for every metadata table.

Feel free to open this conversation back up if you have any concerns. 👍

Reply:
Coming back to this now. I think that we probably don't want to make the `entries` and `files` tables use `StaticTableScan` because `StaticTableScan` assumes that there will be just one task created. The `entries` table is designed to actually create lots of tasks that can run in parallel and read when `rows` is called, in contrast to the static tables that read all necessary files on the driver and create a task that contains the serialized rows.